lib/block.ml

```diff
@@ -295,105 +296,195 @@ let version_enabled version =
+      match header with
       | Some Header.OCaml -> (
           let env = Ocaml_env.mk env in
           match guess_ocaml_kind contents with
-          | `Code -> Ok (OCaml { env; non_det; errors })
+          | `Code ->
+              Util.Result.errorf "invalid toplevel syntax in toplevel blocks."
           | `Toplevel ->
               check_no_errors errors >>= fun () ->
               Ok (Toplevel { env; non_det }) )
-      | _ -> check_no_errors errors >>= fun () -> Ok (Raw { header }) ) )
+      | Some header -> header_not_allowed ~header ~kind
+      | None -> header_required ~headers:[ Header.OCaml ] ~kind )
+  | { file_inc = Some _; _ } -> label_not_allowed ~label:"file" ~kind
+  | { part = Some _; _ } -> label_not_allowed ~label:"part" ~kind
+
+let mk_include ~config ~header ~errors =
+  let kind = "include" in
+  match config with
+  | { file_inc = Some file_included; part; non_det = None; env = None; _ } -> (
+      check_no_errors errors >>= fun () ->
+      match header with
+      | Some Header.OCaml ->
+          let file_kind = Fk_ocaml { part_included = part } in
+          Ok (Include { file_included; file_kind })
+      | _ -> (
+          match part with
+          | None ->
+              let file_kind = Fk_other { header } in
+              Ok (Include { file_included; file_kind })
+          | Some _ -> label_not_allowed ~label:"part" ~kind:"non-OCaml include"
+          ) )
+  | { file_inc = None; _ } -> label_required ~label:"file" ~kind
+  | { non_det = Some _; _ } ->
+      label_not_allowed ~label:"non-deterministic" ~kind
+  | { env = Some _; _ } -> label_not_allowed ~label:"env" ~kind
+
+let mk_raw ~config ~header ~errors =
+  let kind = "raw" in
+  match config with
+  | { file_inc = None; part = None; env = None; _ } ->
+      check_no_errors errors >>= fun () -> Ok (Raw { header })
+  | { file_inc = Some _; _ } -> label_not_allowed ~label:"file" ~kind
+  | { part = Some _; _ } -> label_not_allowed ~label:"part" ~kind
+  | { env = Some _; _ } -> label_not_allowed ~label:"env" ~kind
+
+(* we don't want to sequentially test for include then raw *)
+let mk_include_or_raw ~config ~header ~errors =
+  match config with
+  | { file_inc = Some _; _ } -> mk_include ~config ~header ~errors
+  | { file_inc = None; _ } -> mk_raw ~config ~header ~errors
+
+let ( ||| ) x y = match x with Ok x -> Ok x | Error _ -> y
+
+let mk ~line ~file ~column ~section ~labels ~legacy_labels ~header ~contents
+    ~errors =
+  let block_kind =
+    get_label (function Block_kind x -> Some x | _ -> None) labels
+  in
+  let config = get_label_config labels in
+  ( match block_kind with
+  | Some OCaml -> mk_ocaml ~config ~header ~contents ~errors
+  | Some Cram -> mk_cram ~config ~header ~errors
+  | Some Toplevel -> mk_toplevel ~config ~header ~contents ~errors
+  | Some Include -> mk_include ~config ~header ~errors
+  | None ->
+      mk_ocaml ~config ~header ~contents ~errors
```

Comment:
I think this should fallback to the old block type inference we used to have. That should ensure we don't break the old behaviour + will perform better and allow better error messages than sequentially trying to interpret it as any of the block types.

In particular if we change the order here it will change the error message and that's highly confusing to users.

Reply:
Agreed, done.